### ISSUE TITLE
keke: use StorepointSpider

### ISSUE THREAD

**Issue by davidhicks** [alltheplaces/alltheplaces#4757]
The current KekeSpider (keke.py) uses a SitemapSpider that seemingly can now be replaced with a single request via the StorepointSpider using application key [158d5359b0129d](https://api.storepoint.co/v1/158d5359b0129d/locations) that is found at https://www.kekes.com/locations

**Comment by Cj-Malone** [comment]
In general my preferred order of spiders is:

1. StructuredDataSpider, this data is specifically published to be scraped, it's as close as we can get to having permission.
2. API spiders, well structured data, but no docs.
3. HTML scrapers with xpaths everywhere, just so fragile.

That's my order, @iandees may order it differently. And people may value the cost/efficiency StructuredDataSpider vs api spiders differently.

Specifically for Keke, it's a mix of 1 and 3. So for me it'd come down to the details, how much data can we get from each, and the perceived stability of those xpaths. I've not looked at the details, but I trust you or someone else to ultimately either rewrite the spider or close this issue.

**Comment by iandees** [comment]
Yep, generally I agree with this, with the caveat that (2) is split into "first party APIs" and "third party APIs". When a first-party API is available, I would prefer that over the data available over third party APIs.

**Comment by Cj-Malone** [comment]
If you mean first/third party APIs as in, who hosts them, I'm not sure I care that much any more. I used to, but it's clear this data is from the brands, it's first party data. Maybe we should document it, I'll do a PR tomorrow to add dataset attributes.

If you mean first/third party as in data. I'm still strongly against it, I don't even think it should be "4" in that list. We shouldn't collect it.

**Comment by iandees** [comment]
I just mean that if a business runs their own location API service and then also sends it to where2getit or something, we should prefer the business's own location API service.

I agree that we should not include data aggregators.

**Comment by Cj-Malone** [comment]
I agree in principle, but in reality it's about maintainability. So I guess I wouldn't replace a first party api with a third party one, but writing a new one, I'd lean towards using the third party one, assuming the data is just as good.

It might have been storepoint, but I think it's storerocket, they actually advertise to potential customers that they'll use the data directly from your Google sheet, so brands can avoid duplicate datasets.

We're probably splitting hairs anyway because how many brands have multiple apis? I expect it's not a lot.

**Comment by iandees** [comment]
> how many brands have multiple apis? I expect it's not a lot

Yep, agree.

**Comment by davidhicks** [comment]
I've seen fairly commonly that a brand will have more than one of the following sources of store location data:

1. A store locator page that uses an API (either custom/self-hosted or external SaaS store finder API) to show locations on a map or in a list, generally after a user has searched for a specific city/postcode.
2. A store locator page that links to an index page (HTML) with all stores in a country/region listed with some basic details, including a link to follow to a page specific to each store.
3. A sitemap.xml document that has links to specific store pages.
4. A different API used for the shopping cart for selecting a "click and collect" location for pick-up of items.
5. A different API used for checking stock inventory levels, for example, listing 5 closest stores to a city/postcode with a particular item available.

My take on preferred sources, from highest to lowest preference:
1. If there is only one source with complete data available, then use it. For example, if individual store pages are the only source of opening hours, then a SitemapSpider/CrawlSpider would be preferred.
2. Use the source which is either on the official store locator page, or part of the inventory/cart system, and results in the fewest requests. If there is an API available that can return all store details as a single JSON object, then this approach is most efficient for both alltheplaces users and the brand. For alltheplaces, spiders execute quicker and have less risk of being blocked. For brands, there is lesser load on their servers.
3. Use a third party source, for example, if a retailer always stocks a particular clothing brand, and the clothing brand has their own "which stores stock our products" locator, then this third party locator may have store location data for the retailer.

**Comment by CloCkWeRX** [comment]
Not sure if they are using storepoint anymore
